### PR TITLE
Implement Feign Death resistance

### PR DESCRIPTION
**Pull request by GuybrushGit**
Feign Death is kind of a unique spell, its an aura you apply on yourself but its resistance depends on the mobs you're in combat with, if only one them fails the check the spell resists for all mobs involved.

Follows magic hit formulas: 
http://web.archive.org/web/20061230092631/http://www.thottbot.com:80/?sp=5384

### CODE REVIEW

**Comment by ratkosrb** [comment]
Instead of hard coding the spell Id, you should check if the spell applies aura type 66.

**Comment by ratkosrb** [comment]
Also there is an opcode named SMSG_FEIGN_DEATH_RESISTED. I think that should be sent on resist?

**Comment by GuybrushGit** [comment]
Nice find. Added the resist message and moved the check to a better place.

**Comment by Crusaderz** [comment]
@GuybrushGit 
1. According to your link resist chance is also based on the distance between mob and player.

>From experience, it appears that anything nearby gets a chance to resist, and anything far away does not
>For example, I have seen FD resisted when used out of combat - because there were a number of mobs nearby.  Conversely, I have never seen FD fail, no matter how many mobs are angry at the hunter, if all the mobs are far away (raid pulling). 

We don't know how distance affects FD chance so it seems like you have to improvise. 

2. Will this fix make FD resistable in pvp? Players cannot resist FD since patch 1.7

http://wowwiki.wikia.com/wiki/Patch_1.7.0
>Feign Death is no longer resisted by players.

Thanks.

**Comment by GuybrushGit** [comment]
I read that comment before. Its probably not the distance that affects it but the amount of mobs within a certain range. Lots of mobs = lots of resist checks and if it fails even one the spell resists.

I included a check to prevent it resisting on players.

**Comment by Crusaderz** [comment]
> Its probably not the distance that affects it but the amount of mobs within a certain range.

Well the guy clearly stated "I have never seen FD fail, no matter how many mobs are angry at the hunter, if all the mobs are far away (raid pulling)".

Another source, October 25, 2006
http://wowwiki.wikia.com/wiki/Feign_Death?oldid=300100 

>Feign Death causes the hunter to appear dead. The Hunter loses all aggro. **Nearby** enemies have a chance based on your relative level to resist this effect.

Seems like resist should be range based after all. Implementing resist without range check would lead to 
probems vs raid bosses. Aggro wipe or fd->drink (legit vanilla feature) and raid bosses are higher level so there will be resists.

I also have a question. Will player pets resist FD or your check for players also prevents it?

**Comment by Crusaderz** [comment]
One more.

sep 30 2006
http://wow.allakhazam.com/forum.html?forum=243&mid=115946480249329684#7

>On thing to keep in mind is that FD seems to take proximity to the mob into consideration. If you are a good distance away and FD, you have a greater chance of it being successful. If you are closer to the mob your chances go down slightly.

File: src/game/Spells/SpellAuras.cpp [hunk 1]
```diff
@@ -3253,12 +3253,50 @@ void Aura::HandleFeignDeath(bool apply, bool Real)
...
+                plr->SendFeignDeathResisted();
+                plr->SendAttackSwingCancelAttack();
+            }
+
+            // prevent interrupt message
+            if (GetCasterGuid() == pTarget->GetObjectGuid())
+                pTarget->FinishSpell(CURRENT_GENERIC_SPELL, false);
+            pTarget->InterruptNonMeleeSpells(true);
+
+            GetHolder()->SetAuraDuration(0);
```
Comment (bladezLH) [inline src/game/Spells/SpellAuras.cpp]: I don't think you need to remove the aura or make it end - if it resists they should stay in FD but get wrecked by the mobs iirc. SetFeignDeath also handles calls like InterruptNonMeleeSpells as well. I think you should leave that logic in SetFeignDeath but add another "resist" param and selectively ignore certain functionality (such as setting death state, combatstop, etc)

Comment (GuybrushGit) [reply src/game/Spells/SpellAuras.cpp]: Alright changed it.

**Comment by ratkosrb** [comment]
So are we adding a range check? The creature aggro radius sounded reasonable to me.

**Comment by GuybrushGit** [comment]
Can add, but have no idea what range to use.

**Comment by brotalnia** [comment]
```
refTarget->GetAttackDistance(pTarget)
```

**Comment by GuybrushGit** [comment]
So agro distance then? Feign resists will almost never happen in a raid environment then, since hunters are usually at max range.

**Comment by brotalnia** [comment]
Well the comments seem to support that:
```
Personally, I dont think I have ever had a resist FD to mob/boss if I simply get out of range. It is not that hard to just back up. But saying that, there are times when that is a bit hard to do, e.g. Satura fight. I do not have have Imp FD but have thought about looking into it as well. 
```

http://wow.allakhazam.com/forum.html?forum=243&mid=115946480249329684#7

**Comment by GuybrushGit** [comment]
Added it.

**Pull request merged**